File: tell-dont-ask-kata-kotlin/src/main/java/it/gabrieletondi/telldontaskkata/useCase/OrderCreationUseCase.kt
package it.gabrieletondi.telldontaskkata.useCase


import it.gabrieletondi.telldontaskkata.domain.Order
import it.gabrieletondi.telldontaskkata.domain.OrderItem
import it.gabrieletondi.telldontaskkata.domain.OrderStatus
import it.gabrieletondi.telldontaskkata.domain.Product
import it.gabrieletondi.telldontaskkata.repository.OrderRepository
import it.gabrieletondi.telldontaskkata.repository.ProductCatalog
import java.math.BigDecimal
import java.math.BigDecimal.valueOf
import java.math.RoundingMode.HALF_UP
import java.util.ArrayList

class OrderCreationUseCase(private val orderRepository: OrderRepository, private val productCatalog: ProductCatalog) {

    fun run(request: SellItemsRequest) {
        val order = Order()
        order.status = OrderStatus.CREATED
        order.items =ArrayList()
        order.currency = "EUR"
        order.total = BigDecimal("0.00")
        order.tax = BigDecimal("0.00")

        for (itemRequest in request.requests!!) {
            val product = getProduct(itemRequest)
            val orderItem = OrderItem()
            orderItem.product = product
            orderItem.quantity = itemRequest.quantity
            orderItem.tax = product.price!!.divide(valueOf(100))
                .multiply(product.category!!.taxPercentage!!).setScale(2, HALF_UP).multiply(BigDecimal.valueOf(itemRequest.quantity.toLong()))
            orderItem.taxedAmount =
                product.price!!.add(
                    product.price!!.divide(valueOf(100))
                        .multiply(product.category!!.taxPercentage!!).setScale(2, HALF_UP)
                ).setScale(2, HALF_UP).multiply(BigDecimal.valueOf(itemRequest.quantity.toLong())).setScale(2, HALF_UP)

            order.items!!.add(orderItem)
            order.total = order.total!!.add(orderItem.taxedAmount!!)
            order.tax = order.tax!!.add(orderItem.tax!!)
        }

        orderRepository.save(order)
    }

    private fun getProduct(itemRequest: SellItemRequest): Product {

        return productCatalog.getByName(itemRequest.productName!!) ?: throw UnknownProductException()
    }
}

File: tell-dont-ask-kata-ruby/use_cases/OrderCreationUseCase.rb
require_relative "UnknownProductError"
require_relative "../models/Order"

class OrderCreationUseCase

    attr_accessor :order_database, :product_catalog

    def initialize(order_database:, product_catalog:)
        @order_database = order_database
        @product_catalog = product_catalog
    end

    def run(request:)
        order = Order.new(
          :status => OrderStatus::CREATED,
          :items => [],
          :currency => 'EUR',
          :total => 0.00,
          :tax => 0.00
        )

        request.requests.each do | item_request |
            product = @product_catalog.find_by_name(:name => item_request.product_name)

            if (product.nil?)
                raise UnknownProductError.new()
            else
                unitary_tax = (product.price / 100.0 * product.category.tax_percentage).round(2)
                unitary_taxed_amount = (product.price + unitary_tax).round(2)
                taxed_amount = (unitary_taxed_amount * item_request.quantity).round(2)
                tax_amount = (unitary_tax * item_request.quantity)

                order_item = OrderItem.new(
                    :product => product,
                    :quantity => item_request.quantity,
                    :tax => tax_amount,
                    :taxed_amount => taxed_amount,
                )

                order.items << order_item
                order.total += taxed_amount
                order.tax += tax_amount
            end
        end

        @order_database.save(:order => order)
    end
end